File: day2/day2.py
import re
from pathlib import Path
from functools import reduce
from aoc_utils import timing

# Part one


def parse_line(line: str) -> tuple[str, list[dict[str, int]]]:
    id_line, value_line = line.split(":")
    game_id = int(re.findall(r"Game (\d+)", id_line)[0])
    game_results = []
    for game in value_line.split(";"):
        extraction_results = dict()
        for entry in game.split(","):
            value, color = re.findall(r"(\d+) ([a-z]+)", entry)[0]
            extraction_results[color] = int(value)
        game_results.append(extraction_results)
    return game_id, game_results


def is_game_valid(
    game_results: list[dict[str, int]], value_limits: dict[str, int]
) -> bool:
    return all(map(lambda x: is_extraction_valid(x, value_limits), game_results))


def is_extraction_valid(extraction_results: dict[str, int], value_limits) -> bool:
    return all(value <= value_limits[key] for key, value in extraction_results.items())


with timing():
    total = 0
    value_limits = {"red": 12, "green": 13, "blue": 14}
    with open(Path(__file__).parent / "input.txt") as fin:
        for line in fin.readlines():
            game_id, game_result = parse_line(line)
            if is_game_valid(game_result, value_limits):
                total += game_id
print(total)


# Part two


def find_game_power(game_results: list[dict[str, int]]) -> int:
    mins = {"red": 0, "green": 0, "blue": 0}
    for extraction in game_results:
        for color, value in extraction.items():
            mins[color] = max(mins[color], value)
    return reduce(lambda a, b: a * b, mins.values())


with timing():
    with open(Path(__file__).parent / "input.txt") as fin:
        total = sum(
            map(lambda x: find_game_power(x[-1]), map(parse_line, fin.readlines()))
        )

print(total)

File: day2/day2.jl
function parse_line(line::String, pattern::Regex)::Dict{String,Int}
    values = Dict{String,Int}()
    _, extraction_lines = split(line, ":")
    for extraction_line in split(extraction_lines, ";")
        for color_extraction in split(extraction_line, ",")
            value, color = match(pattern, color_extraction).captures
            value = parse(Int, value)
            if value > get!(values, color, 0)
                values[color] = value
            end
        end
    end
    return values
end


function find_valid(input_path::String)::Int
    pattern = Regex("([0-9]+) (green|blue|red)")
    allowed_maxima = Dict{String,Int}(
        "red" => 12, "green" => 13, "blue" => 14
    )
    maxima = map(x -> parse_line(x, pattern), eachline(input_path))
    valid = map(x -> all(key -> x[key] <= allowed_maxima[key], collect(keys(x))), maxima)
    return sum(map(val -> val[1] * val[2], enumerate(valid)))
end

function find_power(input_path::String)::Int
    pattern = Regex("([0-9]+) (green|blue|red)")
    maxima = map(x -> parse_line(x, pattern), eachline(input_path))
    return sum(map(x -> reduce((a, b) -> a * b, collect(values(x))), maxima))
end

@time result = find_valid(joinpath(@__DIR__, "input.txt"))
println(result)

@time result = find_power(joinpath(@__DIR__, "input.txt"))
println(result)